File: app/src/main/java/com/example/bloodbank/ui/fragment/homeCycle/NotificationSettingsFragment.java
package com.example.bloodbank.ui.fragment.homeCycle;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.bloodbank.R;
import com.example.bloodbank.adapter.CheckBoxAdapter;
import com.example.bloodbank.data.model.General.GenaralResponse;
import com.example.bloodbank.data.model.General.GeneralData;
import com.example.bloodbank.data.model.notificationSettings.NotificationSettings;
import com.example.bloodbank.ui.fragment.BaseFragment;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static com.example.bloodbank.data.api.ApiClient.getClient;
import static com.example.bloodbank.data.local.SharedPreferencesManger.API_TOKEN;
import static com.example.bloodbank.data.local.SharedPreferencesManger.LoadData;
import static com.example.bloodbank.helper.HelperMethods.dismissProgressDialog;
import static com.example.bloodbank.helper.HelperMethods.showProgressDialog;
import static com.example.bloodbank.helper.HelperMethods.showToast;

public class NotificationSettingsFragment extends BaseFragment {
    @BindView(R.id.fragment_notifi_sett_rv_blood_type)
    RecyclerView fragmentNotifiSettRvBloodType;
    @BindView(R.id.fragment_notifi_sett_rv_governorate)
    RecyclerView fragmentNotifiSettRvGovernorate;
    @BindView(R.id.fragment_notifi_sett_cv_blood_type)
    CardView fragmentNotifiSettCvBloodType;
    @BindView(R.id.fragment_notifi_sett_cv_governorate)
    CardView fragmentNotifiSettCvGovernorate;
    @BindView(R.id.fragment_notifi_sett_iv_img_blood_type)
    ImageView fragmentNotifiSettIvImgBloodType;
    @BindView(R.id.fragment_notifi_sett_iv_img_governorate)
    ImageView fragmentNotifiSettIvImgGovernorate;

    private CheckBoxAdapter bloodTypeAdapter, governorateAdapter;
    private List<String> oldBloodTypeList, oldGovernorateList;

    private static final String TAG = "NotificationSettingsFra";

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        setupAcvtivity();
        View view = inflater.inflate(R.layout.fragment_notification_settings, container, false);
        ButterKnife.bind(this, view);

        GridLayoutManager gridLayoutManager = new GridLayoutManager(getActivity(), 3);
        GridLayoutManager gridLayoutManager1 = new GridLayoutManager(getActivity(), 3);
        fragmentNotifiSettRvBloodType.setLayoutManager(gridLayoutManager);
        fragmentNotifiSettRvGovernorate.setLayoutManager(gridLayoutManager1);

        getNotificationSettings();
        return view;
    }


    private void getBloodType() {
        getClient().getbloodTypes().enqueue(new Callback<GenaralResponse>() {
            @Override
            public void onResponse(Call<GenaralResponse> call, Response<GenaralResponse> response) {
                if (response.body().getStatus() == 1) {
                    bloodTypeAdapter = new CheckBoxAdapter(getActivity(), response.body().getData(), oldBloodTypeList);
                    fragmentNotifiSettRvBloodType.setAdapter(bloodTypeAdapter);
                }
            }

            @Override
            public void onFailure(Call<GenaralResponse> call, Throwable t) {

            }
        });
    }

    private void getGovernorates() {
        getClient().getGovernorates().enqueue(new Callback<GenaralResponse>() {
            @Override
            public void onResponse(Call<GenaralResponse> call, Response<GenaralResponse> response) {
                if (response.body().getStatus() == 1) {
                    governorateAdapter = new CheckBoxAdapter(getActivity(), response.body().getData(), oldGovernorateList);
                    fragmentNotifiSettRvGovernorate.setAdapter(governorateAdapter);
                }
            }

            @Override
            public void onFailure(Call<GenaralResponse> call, Throwable t) {

            }
        });
    }

    private void getNotificationSettings() {
        oldBloodTypeList = new ArrayList<>();
        oldGovernorateList = new ArrayList<>();
        showProgressDialog(getActivity(), getString(R.string.please_wait));
        getClient().getNotificationSettings(LoadData(getActivity(), API_TOKEN)).enqueue(new Callback<NotificationSettings>() {
            @Override
            public void onResponse(Call<NotificationSettings> call, Response<NotificationSettings> response) {
                if (response.body().getStatus() == 1) {
                    dismissProgressDialog();
                    try {
                        oldBloodTypeList.addAll(response.body().getData().getBloodTypes());
                        oldGovernorateList.addAll(response.body().getData().getGovernorates());
                        getBloodType();
                        getGovernorates();

                        Log.i(TAG, " old " + oldBloodTypeList);
                    } catch (Exception e) {
                        Log.i(TAG, "onResponse: " + e.getMessage());
                    }
                }
            }

            @Override
            public void onFailure(Call<NotificationSettings> call, Throwable t) {

            }
        });
    }

    private void changeNotificationSettings() {
        showProgressDialog(getActivity(), getString(R.string.please_wait));
        getClient().changeNotificationSettings(LoadData(getActivity(), API_TOKEN), governorateAdapter.newIds, bloodTypeAdapter.newIds).enqueue(new Callback<NotificationSettings>() {
            @Override
            public void onResponse(Call<NotificationSettings> call, Response<NotificationSettings> response) {
                dismissProgressDialog();
                if (response.body().getStatus() == 1) {
                    showToast(getActivity(), response.body().getMsg());
                }
            }

            @Override
            public void onFailure(Call<NotificationSettings> call, Throwable t) {

            }
        });
    }

    private void visible(View view, ImageView imageView) {
        if (view.getVisibility() == View.GONE) {
            view.setVisibility(View.VISIBLE);
            imageView.setImageResource(R.drawable.ic_remove);
        } else {
            view.setVisibility(View.GONE);
            imageView.setImageResource(R.drawable.ic_add);
        }
    }

    @OnClick({R.id.fragment_notifi_sett_btn_show_blood_type, R.id.fragment_notifi_sett_btn_show_governorate, R.id.fragment_notifi_sett_btn_save})
    public void onViewClicked(View view) {
        switch (view.getId()) {
            case R.id.fragment_notifi_sett_btn_show_blood_type:
                visible(fragmentNotifiSettCvBloodType, fragmentNotifiSettIvImgBloodType);
                break;
            case R.id.fragment_notifi_sett_btn_show_governorate:
                visible(fragmentNotifiSettCvGovernorate, fragmentNotifiSettIvImgGovernorate);
                break;
            case R.id.fragment_notifi_sett_btn_save:
                changeNotificationSettings();
                break;
        }
    }

    @Override
    public void onBack() {
        super.onBack();
    }
}

File: app/src/main/java/com/example/bloodbank/ui/fragment/homeCycle/notifications/NotificationSettingsFragment.kt
package com.example.bloodbank.ui.fragment.homeCycle.notifications

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.GridLayoutManager
import com.example.bloodbank.R
import com.example.bloodbank.data.api.ApiClient.client
import com.example.bloodbank.data.local.SharedPreferencesManger.LoadData
import com.example.bloodbank.data.model.general.GeneralData
import com.example.bloodbank.data.model.general.GeneralResponse
import com.example.bloodbank.data.model.notificationSettings.NotificationSettings
import com.example.bloodbank.databinding.FragmentNotificationSettingsBinding
import com.example.bloodbank.extensions.inflateWithBinding
import com.example.bloodbank.extensions.visible
import com.example.bloodbank.helper.API_TOKEN
import com.example.bloodbank.helper.HelperMethods.dismissProgressDialog
import com.example.bloodbank.helper.HelperMethods.showProgressDialog
import com.example.bloodbank.helper.HelperMethods.showToast
import com.example.bloodbank.ui.adapter.CheckBoxAdapter
import com.example.bloodbank.ui.fragment.BaseFragment
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class NotificationSettingsFragment : BaseFragment(), View.OnClickListener {

    private var bloodTypeAdapter: CheckBoxAdapter? = null
    private var governorateAdapter: CheckBoxAdapter? = null
    private lateinit var oldBloodTypeList: MutableList<String>
    private lateinit var oldGovernorateList: MutableList<String>

    lateinit var binding: FragmentNotificationSettingsBinding

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        setupActivity()
        binding = inflater.inflateWithBinding(R.layout.fragment_notification_settings, container)

        val gridLayoutManager = GridLayoutManager(activity, 3)
        val gridLayoutManager1 = GridLayoutManager(activity, 3)
        binding.fragmentNotifiSettRvBloodType.layoutManager = gridLayoutManager
        binding.fragmentNotifiSettRvGovernorate.layoutManager = gridLayoutManager1
        notificationSettings()
        binding.fragmentNotifiSettBtnShowBloodType.setOnClickListener(this)
        binding.fragmentNotifiSettBtnShowGovernorate.setOnClickListener(this)
        binding.fragmentNotifiSettBtnSave.setOnClickListener(this)

        return binding.root
    }

    fun bloodType() {
        client.getbloodTypes().enqueue(object : Callback<GeneralResponse> {
            override fun onResponse(call: Call<GeneralResponse>, response: Response<GeneralResponse>) {
                if (response.body()!!.status == 1) {
                    bloodTypeAdapter = CheckBoxAdapter((response.body()!!.data as MutableList<GeneralData>?)!!, oldBloodTypeList)
                    binding.fragmentNotifiSettRvBloodType.adapter = bloodTypeAdapter
                }
            }

            override fun onFailure(call: Call<GeneralResponse>, t: Throwable) {}
        })
    }

    fun governorates() {
        client.governorates.enqueue(object : Callback<GeneralResponse> {
            override fun onResponse(call: Call<GeneralResponse>, response: Response<GeneralResponse>) {
                if (response.body()!!.status == 1) {
                    governorateAdapter = CheckBoxAdapter((response.body()!!.data as MutableList<GeneralData>?)!!, oldGovernorateList)
                    binding.fragmentNotifiSettRvGovernorate.adapter = governorateAdapter
                }
            }

            override fun onFailure(call: Call<GeneralResponse>, t: Throwable) {}
        })
    }

    private fun notificationSettings() {
        oldBloodTypeList = mutableListOf()
        oldGovernorateList = mutableListOf()
        showProgressDialog(activity, getString(R.string.please_wait))
        client.getNotificationSettings(LoadData(activity!!, API_TOKEN)!!).enqueue(object : Callback<NotificationSettings> {
            override fun onResponse(call: Call<NotificationSettings>, response: Response<NotificationSettings>) {
                if (response.body()!!.status == 1) {
                    dismissProgressDialog()
                    try {
                        oldBloodTypeList.addAll(response.body()!!.data!!.bloodTypes!!)
                        oldGovernorateList.addAll(response.body()!!.data!!.governorates!!)
                        bloodType()
                        governorates()
                        Log.i(TAG, " old $oldBloodTypeList")
                    } catch (e: Exception) {
                        Log.i(TAG, "onResponse: " + e.message)
                    }
                }
            }

            override fun onFailure(call: Call<NotificationSettings>, t: Throwable) {}
        })
    }

    private fun changeNotificationSettings() {
        showProgressDialog(activity, getString(R.string.please_wait))
        client.changeNotificationSettings(LoadData(activity!!, API_TOKEN)!!, governorateAdapter!!.newIds, bloodTypeAdapter!!.newIds).enqueue(object : Callback<NotificationSettings> {
            override fun onResponse(call: Call<NotificationSettings>, response: Response<NotificationSettings>) {
                dismissProgressDialog()
                if (response.body()!!.status == 1) {
                    showToast(activity, response.body()!!.msg)
                }
            }

            override fun onFailure(call: Call<NotificationSettings>, t: Throwable) {}
        })
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.fragment_notifi_sett_btn_show_blood_type -> binding.fragmentNotifiSettCvBloodType.visible(binding.fragmentNotifiSettIvImgBloodType)
            R.id.fragment_notifi_sett_btn_show_governorate -> binding.fragmentNotifiSettCvGovernorate.visible(binding.fragmentNotifiSettIvImgGovernorate)
            R.id.fragment_notifi_sett_btn_save -> changeNotificationSettings()
        }
    }

    companion object {
        private const val TAG = "NotificationSettingsFra"
    }
}